xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java

```diff
@@ -498,60 +590,123 @@ private <T extends ResourceUpdate> void handleResourceUpdate(
       // For State of the World services, notify watchers when their watched resource is missing
       // from the ADS update. Note that we can only do this if the resource update is coming from
       // the same xDS server that the ResourceSubscriber is subscribed to.
-      if (subscriber.serverInfo.equals(args.serverInfo)) {
-        subscriber.onAbsent(processingTracker);
+      if (getActiveCpc(subscriber.authority) == controlPlaneClient) {
+        subscriber.onAbsent(processingTracker, args.serverInfo);
+      }
+    }
+  }
+
+  private void shutdownLowerPriorityCpcs(ControlPlaneClient activatedCpc) {
+    // For each authority, remove any control plane clients, with lower priority than the activated
+    // one, from activatedCpClients storing them all in cpcsToShutdown.
+    Set<ControlPlaneClient> cpcsToShutdown = new HashSet<>();
+    for ( List<ControlPlaneClient> cpcsForAuth : activatedCpClients.values()) {
+      if (cpcsForAuth == null) {
+        continue;
+      }
+      int index = cpcsForAuth.indexOf(activatedCpc);
+      if (index > -1) {
+        cpcsToShutdown.addAll(cpcsForAuth.subList(index + 1, cpcsForAuth.size()));
+        cpcsForAuth.subList(index + 1, cpcsForAuth.size()).clear(); // remove lower priority cpcs
+      }
+    }
+
+    // Shutdown any lower priority control plane clients identified above that aren't still being
+    // used by another authority.  If they are still being used let the XDS server know that we
+    // no longer are interested in subscriptions for authorities we are no longer responsible for.
+    for (ControlPlaneClient cpc : cpcsToShutdown) {
+      if (activatedCpClients.values().stream().noneMatch(list -> list.contains(cpc))) {
+        cpc.shutdown();
+        serverCpClientMap.remove(cpc.getServerInfo());
+      } else {
+        cpc.sendDiscoveryRequests();
+      }
+    }
+  }
+
+  private boolean fallBackToCpc(ControlPlaneClient fallbackCpc, String authority,
+                                ControlPlaneClient oldCpc) {
+    boolean didFallback = false;
+    if (fallbackCpc != null && ! fallbackCpc.isInError()) {
+      logger.log(XdsLogLevel.INFO, "Falling back to XDS server {0}",
+          fallbackCpc.getServerInfo().target());
+
+      // Get authorities that aren't falling back
+      // If we don't already have a cached LDS resource, cache the current data value
+
+      restartMatchingSubscriberTimers(authority);
```

Comment:
Hmmm... It's not clear to me what the timeout behavior should be. Starting/restarting these timers now means the can start before onReady() on the current cpc. We could _stop_ the timers here and have them be started in onReady() like normal. But also, if it takes over 20 seconds for the cpc to fail without an initial response, then these resources will be considered does-not-exist, at which point we have cached results and we won't fallback further. The gRFC doesn't talk about the interplay at all...

Reply:
Since we are falling back, there could be existing timers that we definitely don't want.  At a minimum we should cancel those timers.  It feels to me like the right approach is to start subscriber timers for any resource that doesn't have one (or a result) when we are sending the discovery request.  Probably best to add a `ResourceStore.startMissingResourceTimers(Collection<String> resourceNames, XdsResourceType resourceType)`.

Reply:
Timers previously were only running between the stream's first onReady() and the stream's end. I think we should continue that for the moment. That'd mean delete this `restartmatchingSubscriberTimers()`, and we'd rely on the existing `startSubscriberTimersIfNeeded()` called by `handleStreamRestarted()`.

But I do think we need to discuss this topic with others.

Reply:
It makes sense to let the `handleStreamRestarted` set the timers instead of because we detected a fallback.  Then it's just a question of whether we have `watchXdsResource` starts a timer or have `adjustResourceSubscriptions` call a new `ResourceStore` method.

Reply:
Having `watchXdsResource()` do `if (cpcToUse.cpc.isConnected()) subscriber.restartTimer()` seems behaviorally correct and easy. But I think it isn't the real solution. The timer bug here also exists if we fallback to a `isConnected() == true` cpc... right? Both of those call `adjustResourceSubscriptions()`.

This PR essentially renamed `XdsClient.startSubscriberTimersIfNeeded()` to `XdsResponseHandler.handleStreamRestarted()`. As discussed in our meeting, let's trigger the timer starting from `adjustResourceSubscriptions()` (with some new method to call back into xds client) and then delete `handleStreamRestarted()`.